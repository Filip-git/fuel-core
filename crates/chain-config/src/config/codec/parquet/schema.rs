use std::sync::Arc;

use parquet::{
    basic::Repetition,
    schema::types::Type,
};

use crate::{
    config::{
        contract_balance::ContractBalance,
        contract_state::ContractStateConfig,
    },
    CoinConfig,
    ContractConfig,
    MessageConfig,
};

pub trait Schema {
    fn schema() -> Type;
    fn num_of_columns() -> usize {
        Self::schema().get_fields().len()
    }
}

impl Schema for ContractConfig {
    fn schema() -> Type {
        use parquet::basic::Type as PhysicalType;
        let contract_id = Type::primitive_type_builder(
            "contract_id",
            PhysicalType::FIXED_LEN_BYTE_ARRAY,
        )
        .with_length(32)
        .with_repetition(Repetition::REQUIRED)
        .build()
        .unwrap();

        let code = Type::primitive_type_builder("code", PhysicalType::BYTE_ARRAY)
            .with_repetition(Repetition::REQUIRED)
            .build()
            .unwrap();

        let salt =
            Type::primitive_type_builder("salt", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();

        let tx_id =
            Type::primitive_type_builder("tx_id", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();

        let output_index =
            Type::primitive_type_builder("output_index", PhysicalType::INT32)
                .with_converted_type(parquet::basic::ConvertedType::UINT_8)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();

        let tx_pointer_block_height =
            Type::primitive_type_builder("tx_pointer_block_height", PhysicalType::INT32)
                .with_converted_type(parquet::basic::ConvertedType::UINT_32)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();

        let tx_pointer_tx_idx =
            Type::primitive_type_builder("tx_pointer_tx_idx", PhysicalType::INT32)
                .with_converted_type(parquet::basic::ConvertedType::UINT_16)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();

        parquet::schema::types::Type::group_type_builder("ContractConfig")
            .with_fields(
                [
                    contract_id,
                    code,
                    salt,
                    tx_id,
                    output_index,
                    tx_pointer_block_height,
                    tx_pointer_tx_idx,
                ]
                .map(Arc::new)
                .to_vec(),
            )
            .build()
            .unwrap()
    }
}

impl Schema for MessageConfig {
    fn schema() -> Type {
        use parquet::basic::Type as PhysicalType;
        let sender =
            Type::primitive_type_builder("sender", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();
        let recipient =
            Type::primitive_type_builder("recipient", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();
        let nonce =
            Type::primitive_type_builder("nonce", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();
        let amount = Type::primitive_type_builder("amount", PhysicalType::INT64)
            .with_converted_type(parquet::basic::ConvertedType::UINT_64)
            .with_repetition(Repetition::REQUIRED)
            .build()
            .unwrap();
        let data = Type::primitive_type_builder("data", PhysicalType::BYTE_ARRAY)
            .with_repetition(Repetition::REQUIRED)
            .build()
            .unwrap();
        let da_height = Type::primitive_type_builder("da_height", PhysicalType::INT64)
            .with_converted_type(parquet::basic::ConvertedType::UINT_64)
            .with_repetition(Repetition::REQUIRED)
            .build()
            .unwrap();

        parquet::schema::types::Type::group_type_builder("CoinConfig")
            .with_fields(
                [sender, recipient, nonce, amount, data, da_height]
                    .map(Arc::new)
                    .to_vec(),
            )
            .build()
            .unwrap()
    }
}

impl Schema for CoinConfig {
    fn schema() -> Type {
        use parquet::basic::Type as PhysicalType;
        let tx_id =
            Type::primitive_type_builder("tx_id", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();
        let output_index =
            Type::primitive_type_builder("output_index", PhysicalType::INT32)
                .with_converted_type(parquet::basic::ConvertedType::UINT_8)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();
        let tx_pointer_block_height =
            Type::primitive_type_builder("tx_pointer_block_height", PhysicalType::INT32)
                .with_converted_type(parquet::basic::ConvertedType::UINT_32)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();
        let tx_pointer_tx_idx =
            Type::primitive_type_builder("tx_pointer_tx_idx", PhysicalType::INT32)
                .with_converted_type(parquet::basic::ConvertedType::UINT_16)
                .with_repetition(Repetition::OPTIONAL)
                .build()
                .unwrap();
        let maturity = Type::primitive_type_builder("maturity", PhysicalType::INT32)
            .with_converted_type(parquet::basic::ConvertedType::UINT_32)
            .with_repetition(Repetition::OPTIONAL)
            .build()
            .unwrap();
        let owner =
            Type::primitive_type_builder("owner", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();
        let amount = Type::primitive_type_builder("amount", PhysicalType::INT64)
            .with_converted_type(parquet::basic::ConvertedType::UINT_64)
            .with_repetition(Repetition::REQUIRED)
            .build()
            .unwrap();
        let asset_id =
            Type::primitive_type_builder("asset_id", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();

        parquet::schema::types::Type::group_type_builder("CoinConfig")
            .with_fields(
                [
                    tx_id,
                    output_index,
                    tx_pointer_block_height,
                    tx_pointer_tx_idx,
                    maturity,
                    owner,
                    amount,
                    asset_id,
                ]
                .map(Arc::new)
                .to_vec(),
            )
            .build()
            .unwrap()
    }
}

impl Schema for ContractStateConfig {
    fn schema() -> Type {
        use parquet::basic::Type as PhysicalType;
        let contract_id = Type::primitive_type_builder(
            "contract_id",
            PhysicalType::FIXED_LEN_BYTE_ARRAY,
        )
        .with_length(32)
        .with_repetition(Repetition::REQUIRED)
        .build()
        .unwrap();
        let key = Type::primitive_type_builder("key", PhysicalType::FIXED_LEN_BYTE_ARRAY)
            .with_length(32)
            .with_repetition(Repetition::REQUIRED)
            .build()
            .unwrap();
        let value =
            Type::primitive_type_builder("value", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();

        parquet::schema::types::Type::group_type_builder("ContractStateConfig")
            .with_fields([contract_id, key, value].map(Arc::new).to_vec())
            .build()
            .unwrap()
    }
}

impl Schema for ContractBalance {
    fn schema() -> Type {
        use parquet::basic::Type as PhysicalType;
        let contract_id = Type::primitive_type_builder(
            "contract_id",
            PhysicalType::FIXED_LEN_BYTE_ARRAY,
        )
        .with_length(32)
        .with_repetition(Repetition::REQUIRED)
        .build()
        .unwrap();
        let asset_id =
            Type::primitive_type_builder("asset_id", PhysicalType::FIXED_LEN_BYTE_ARRAY)
                .with_length(32)
                .with_repetition(Repetition::REQUIRED)
                .build()
                .unwrap();
        let amount = Type::primitive_type_builder("amount", PhysicalType::INT64)
            .with_converted_type(parquet::basic::ConvertedType::UINT_64)
            .with_repetition(Repetition::REQUIRED)
            .build()
            .unwrap();

        parquet::schema::types::Type::group_type_builder("ContractBalance")
            .with_fields([contract_id, asset_id, amount].map(Arc::new).to_vec())
            .build()
            .unwrap()
    }
}
