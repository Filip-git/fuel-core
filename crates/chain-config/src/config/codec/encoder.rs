use std::path::{
    Path,
    PathBuf,
};

use ::parquet::basic::ZstdLevel;

use crate::{
    config::{
        contract_balance::ContractBalance,
        contract_state::ContractStateConfig,
    },
    CoinConfig,
    ContractConfig,
    MessageConfig,
    StateConfig,
};

#[cfg(feature = "parquet")]
use crate::config::codec::parquet::Schema;

#[cfg(feature = "parquet")]
use super::parquet;

enum EncoderType {
    Json {
        buffer: StateConfig,
        state_file_path: PathBuf,
    },
    #[cfg(feature = "parquet")]
    Parquet {
        coins: parquet::Encoder<std::fs::File, CoinConfig>,
        messages: parquet::Encoder<std::fs::File, MessageConfig>,
        contracts: parquet::Encoder<std::fs::File, ContractConfig>,
        contract_state: parquet::Encoder<std::fs::File, ContractStateConfig>,
        contract_balance: parquet::Encoder<std::fs::File, ContractBalance>,
    },
}

pub struct Encoder {
    encoder: EncoderType,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
#[cfg(feature = "parquet")]
#[cfg_attr(test, derive(strum::EnumIter))]
pub enum ZstdCompressionLevel {
    Uncompressed,
    Level1,
    Level2,
    Level3,
    Level4,
    Level5,
    Level6,
    Level7,
    Level8,
    Level9,
    Level10,
    Level11,
    Level12,
    Level13,
    Level14,
    Level15,
    Level16,
    Level17,
    Level18,
    Level19,
    Level20,
    Level21,
    Max,
}

#[cfg(feature = "parquet")]
impl TryFrom<u8> for ZstdCompressionLevel {
    type Error = anyhow::Error;
    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(Self::Uncompressed),
            1 => Ok(Self::Level1),
            2 => Ok(Self::Level2),
            3 => Ok(Self::Level3),
            4 => Ok(Self::Level4),
            5 => Ok(Self::Level5),
            6 => Ok(Self::Level6),
            7 => Ok(Self::Level7),
            8 => Ok(Self::Level8),
            9 => Ok(Self::Level9),
            10 => Ok(Self::Level10),
            11 => Ok(Self::Level11),
            12 => Ok(Self::Level12),
            13 => Ok(Self::Level13),
            14 => Ok(Self::Level14),
            15 => Ok(Self::Level15),
            16 => Ok(Self::Level16),
            17 => Ok(Self::Level17),
            18 => Ok(Self::Level18),
            19 => Ok(Self::Level19),
            20 => Ok(Self::Level20),
            21 => Ok(Self::Level21),
            22 => Ok(Self::Max),
            _ => {
                anyhow::bail!("Compression level {value} outside of allowed range 0..=22")
            }
        }
    }
}

#[cfg(feature = "parquet")]
impl From<ZstdCompressionLevel> for u8 {
    fn from(value: ZstdCompressionLevel) -> Self {
        match value {
            ZstdCompressionLevel::Uncompressed => 0,
            ZstdCompressionLevel::Level1 => 1,
            ZstdCompressionLevel::Level2 => 2,
            ZstdCompressionLevel::Level3 => 3,
            ZstdCompressionLevel::Level4 => 4,
            ZstdCompressionLevel::Level5 => 5,
            ZstdCompressionLevel::Level6 => 6,
            ZstdCompressionLevel::Level7 => 7,
            ZstdCompressionLevel::Level8 => 8,
            ZstdCompressionLevel::Level9 => 9,
            ZstdCompressionLevel::Level10 => 10,
            ZstdCompressionLevel::Level11 => 11,
            ZstdCompressionLevel::Level12 => 12,
            ZstdCompressionLevel::Level13 => 13,
            ZstdCompressionLevel::Level14 => 14,
            ZstdCompressionLevel::Level15 => 15,
            ZstdCompressionLevel::Level16 => 16,
            ZstdCompressionLevel::Level17 => 17,
            ZstdCompressionLevel::Level18 => 18,
            ZstdCompressionLevel::Level19 => 19,
            ZstdCompressionLevel::Level20 => 20,
            ZstdCompressionLevel::Level21 => 21,
            ZstdCompressionLevel::Max => 22,
        }
    }
}

#[cfg(feature = "parquet")]
impl From<ZstdCompressionLevel> for ::parquet::basic::Compression {
    fn from(value: ZstdCompressionLevel) -> Self {
        if let ZstdCompressionLevel::Uncompressed = value {
            Self::UNCOMPRESSED
        } else {
            let level = ZstdLevel::try_new(u8::from(value) as i32)
                .expect("our range to mimic the parquet zstd range");
            Self::ZSTD(level)
        }
    }
}

impl Encoder {
    pub fn json(snapshot_dir: impl AsRef<Path>) -> Self {
        Self {
            encoder: EncoderType::Json {
                state_file_path: snapshot_dir.as_ref().join("state.json"),
                buffer: StateConfig::default(),
            },
        }
    }

    #[cfg(feature = "parquet")]
    pub fn parquet(
        snapshot_dir: impl AsRef<Path>,
        compression_level: ZstdCompressionLevel,
    ) -> anyhow::Result<Self> {
        use ::parquet::basic::Compression;
        use std::fs::File;

        fn create_encoder<T>(
            path: &Path,
            name: &str,
            compression: Compression,
        ) -> anyhow::Result<parquet::Encoder<File, T>>
        where
            T: Schema,
        {
            let path = path.join(format!("{name}.parquet"));
            let file = std::fs::File::create(path)?;
            parquet::Encoder::new(file, compression)
        }

        let path = snapshot_dir.as_ref();
        let compression = compression_level.into();

        Ok(Self {
            encoder: EncoderType::Parquet {
                coins: create_encoder(path, "coins", compression)?,
                messages: create_encoder(path, "messages", compression)?,
                contracts: create_encoder(path, "contracts", compression)?,
                contract_state: create_encoder(path, "contract_state", compression)?,
                contract_balance: create_encoder(path, "contract_balance", compression)?,
            },
        })
    }

    pub fn write_coins(&mut self, elements: Vec<CoinConfig>) -> anyhow::Result<()> {
        match &mut self.encoder {
            EncoderType::Json { buffer: state, .. } => {
                state.coins.extend(elements);
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet { coins, .. } => coins.write(elements),
        }
    }

    pub fn write_contracts(
        &mut self,
        elements: Vec<ContractConfig>,
    ) -> anyhow::Result<()> {
        match &mut self.encoder {
            EncoderType::Json { buffer: state, .. } => {
                state.contracts.extend(elements);
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet { contracts, .. } => contracts.write(elements),
        }
    }

    pub fn write_messages(&mut self, elements: Vec<MessageConfig>) -> anyhow::Result<()> {
        match &mut self.encoder {
            EncoderType::Json { buffer: state, .. } => {
                state.messages.extend(elements);
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet { messages, .. } => messages.write(elements),
        }
    }

    pub fn write_contract_state(
        &mut self,
        elements: Vec<ContractStateConfig>,
    ) -> anyhow::Result<()> {
        match &mut self.encoder {
            EncoderType::Json { buffer: state, .. } => {
                state.contract_state.extend(elements);
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet { contract_state, .. } => contract_state.write(elements),
        }
    }

    pub fn write_contract_balance(
        &mut self,
        elements: Vec<ContractBalance>,
    ) -> anyhow::Result<()> {
        match &mut self.encoder {
            EncoderType::Json { buffer: state, .. } => {
                state.contract_balance.extend(elements);
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet {
                contract_balance, ..
            } => contract_balance.write(elements),
        }
    }

    pub fn close(self) -> anyhow::Result<()> {
        match self.encoder {
            EncoderType::Json {
                buffer,
                state_file_path,
            } => {
                let file = std::fs::File::create(state_file_path)?;
                serde_json::to_writer(file, &buffer)?;
                Ok(())
            }
            #[cfg(feature = "parquet")]
            EncoderType::Parquet {
                coins,
                messages,
                contracts,
                contract_state,
                contract_balance,
            } => {
                coins.close()?;
                messages.close()?;
                contracts.close()?;
                contract_state.close()?;
                contract_balance.close()?;
                Ok(())
            }
        }
    }
}

#[cfg(feature = "random")]
#[cfg(test)]
mod tests {
    use super::*;
    use itertools::Itertools;

    #[test]
    fn json_encoder_generates_single_file_with_expected_name() {
        // given
        let dir = tempfile::tempdir().unwrap();
        let encoder = Encoder::json(dir.path());

        // when
        encoder.close().unwrap();

        // then
        let entries: Vec<_> = dir.path().read_dir().unwrap().try_collect().unwrap();

        match entries.as_slice() {
            [entry] => assert_eq!(entry.path(), dir.path().join("state.json")),
            _ => panic!("Expected single file \"state.json\""),
        }
    }

    #[cfg(feature = "parquet")]
    #[test]
    fn parquet_encoder_generates_expected_filenames() {
        // given
        let dir = tempfile::tempdir().unwrap();
        let encoder =
            Encoder::parquet(dir.path(), ZstdCompressionLevel::Uncompressed).unwrap();

        // when
        encoder.close().unwrap();

        // then
        let entries: std::collections::HashSet<_> = dir
            .path()
            .read_dir()
            .unwrap()
            .map_ok(|entry| entry.path())
            .try_collect()
            .unwrap();
        let expected_files = std::collections::HashSet::from(
            [
                "coins.parquet",
                "messages.parquet",
                "contracts.parquet",
                "contract_state.parquet",
                "contract_balance.parquet",
            ]
            .map(|name| dir.path().join(name)),
        );

        assert_eq!(entries, expected_files);
    }

    #[cfg(feature = "parquet")]
    #[test]
    fn parquet_encoder_encodes_types_in_expected_files() {
        use rand::SeedableRng;
        let mut rng = rand::rngs::StdRng::seed_from_u64(0);

        test_data_written_in_expected_file::<CoinConfig>(
            &mut rng,
            "coins.parquet",
            |coins, encoder| encoder.write_coins(coins),
        );

        test_data_written_in_expected_file::<MessageConfig>(
            &mut rng,
            "messages.parquet",
            |coins, encoder| encoder.write_messages(coins),
        );

        test_data_written_in_expected_file::<ContractConfig>(
            &mut rng,
            "contracts.parquet",
            |coins, encoder| encoder.write_contracts(coins),
        );

        test_data_written_in_expected_file::<ContractStateConfig>(
            &mut rng,
            "contract_state.parquet",
            |coins, encoder| encoder.write_contract_state(coins),
        );

        test_data_written_in_expected_file::<ContractBalance>(
            &mut rng,
            "contract_balance.parquet",
            |coins, encoder| encoder.write_contract_balance(coins),
        );
    }

    #[cfg(feature = "parquet")]
    fn test_data_written_in_expected_file<T>(
        rng: impl rand::Rng,
        expected_filename: &str,
        write: impl FnOnce(Vec<T>, &mut Encoder) -> anyhow::Result<()>,
    ) where
        parquet::Decoder<std::fs::File, T>:
            Iterator<Item = anyhow::Result<crate::Group<T>>>,
        T: crate::Randomize + PartialEq + ::core::fmt::Debug + Clone,
    {
        use std::fs::File;
        // given
        let dir = tempfile::tempdir().unwrap();
        let mut encoder =
            Encoder::parquet(dir.path(), ZstdCompressionLevel::Uncompressed).unwrap();
        let original_data = vec![T::randomize(rng)];

        // when
        write(original_data.clone(), &mut encoder).unwrap();
        encoder.close().unwrap();

        // then
        let file = std::fs::File::open(dir.path().join(expected_filename)).unwrap();
        let decoded = parquet::Decoder::<File, T>::new(file)
            .unwrap()
            .collect::<Result<Vec<_>, _>>()
            .unwrap();
        assert_eq!(original_data, decoded.first().unwrap().data);
    }

    #[cfg(feature = "parquet")]
    #[test]
    fn all_compressions_are_valid() {
        use ::parquet::basic::Compression;
        use strum::IntoEnumIterator;
        for level in ZstdCompressionLevel::iter() {
            let _ = Compression::from(level);
        }
    }
}
