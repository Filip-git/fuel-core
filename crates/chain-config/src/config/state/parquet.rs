mod decode;
mod encode;
mod schema;

pub use decode::*;
pub use encode::*;
pub use schema::*;
