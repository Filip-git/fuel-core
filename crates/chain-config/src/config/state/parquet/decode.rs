use std::marker::PhantomData;

use anyhow::{
    anyhow,
    bail,
    Context,
};
use fuel_core_types::{
    blockchain::primitives::DaBlockHeight,
    fuel_types::{
        Address,
        AssetId,
        BlockHeight,
        Bytes32,
        ContractId,
        Nonce,
    },
    fuel_vm::Salt,
};
use parquet::{
    file::{
        reader::{
            ChunkReader,
            FileReader,
        },
        serialized_reader::SerializedFileReader,
    },
    record::{
        Field,
        Row,
    },
};

use crate::{
    config::{
        contract_balance::ContractBalanceConfig,
        contract_state::ContractStateConfig,
        state::{
            Group,
            GroupResult,
        },
    },
    CoinConfig,
    ContractConfig,
    MessageConfig,
};

pub struct Decoder<R: ChunkReader, T> {
    data_source: SerializedFileReader<R>,
    group_index: usize,
    _data: PhantomData<T>,
}

impl<R, T> Decoder<R, T>
where
    R: ChunkReader + 'static,
    T: TryFrom<Row, Error = anyhow::Error>,
{
    fn current_group(&self) -> anyhow::Result<Group<T>> {
        let data = self
            .data_source
            .get_row_group(self.group_index)?
            .get_row_iter(None)?
            .map(|result| {
                result
                    .map_err(|e| anyhow!(e))
                    .and_then(|row| T::try_from(row))
            })
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Group {
            index: self.group_index,
            data,
        })
    }
}

impl<R, T> Iterator for Decoder<R, T>
where
    R: ChunkReader + 'static,
    T: TryFrom<Row, Error = anyhow::Error>,
{
    type Item = GroupResult<T>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.group_index >= self.data_source.metadata().num_row_groups() {
            return None;
        }

        let group = self.current_group();
        self.group_index = self.group_index.saturating_add(1);

        Some(group)
    }

    fn nth(&mut self, n: usize) -> Option<Self::Item> {
        self.group_index = self.group_index.saturating_add(n);
        self.next()
    }
}

impl<R: ChunkReader + 'static, T> Decoder<R, T> {
    pub fn new(reader: R) -> anyhow::Result<Self> {
        Ok(Self {
            data_source: SerializedFileReader::new(reader)?,
            group_index: 0,
            _data: PhantomData,
        })
    }
}

impl TryFrom<Row> for CoinConfig {
    type Error = anyhow::Error;

    fn try_from(row: Row) -> Result<Self, Self::Error> {
        let mut iter = row.get_column_iter();
        let mut next_field = || {
            iter.next()
                .map(|el| el.1)
                .ok_or_else(|| anyhow!("Expected at least one more field. Row: {row:?}"))
        };

        let tx_id = next_field()
            .and_then(decode_as_optional_bytes_32)
            .context("While decoding `tx_id`")?;

        let output_index = next_field()
            .and_then(decode_as_optional_u8)
            .context("While decoding `output_index`")?;

        let tx_pointer_block_height = next_field()
            .and_then(decode_as_optional_block_height)
            .context("While decoding `tx_pointer_block_height`")?;

        let tx_pointer_tx_idx = next_field()
            .and_then(decode_as_optional_u16)
            .context("While decoding `tx_pointer_tx_idx`")?;

        let maturity = next_field()
            .and_then(decode_as_optional_block_height)
            .context("While decoding `maturiy`")?;

        let owner = next_field()
            .and_then(decode_as_address)
            .context("While decoding `owner`")?;

        let amount = next_field()
            .and_then(decode_as_u64)
            .context("While decoding `amount`")?;

        let asset_id = next_field()
            .and_then(decode_as_asset_id)
            .context("While decoding `assert_id`")?;

        Ok(Self {
            tx_id,
            output_index,
            tx_pointer_block_height,
            tx_pointer_tx_idx,
            maturity,
            owner,
            amount,
            asset_id,
        })
    }
}

fn decode_as_bytes(field: &Field) -> anyhow::Result<&[u8]> {
    decode_as_optional_bytes(field)?.ok_or_else(|| anyhow!("Cannot be NULL"))
}

fn decode_as_optional_bytes(field: &Field) -> anyhow::Result<Option<&[u8]>> {
    let bytes = match field {
        Field::Bytes(bytes) => Some(bytes.data()),
        Field::Null => None,
        field => bail!("Unexpected field: '{field}'"),
    };
    Ok(bytes)
}

fn decode_as_optional_u8(field: &Field) -> anyhow::Result<Option<u8>> {
    let bytes = match field {
        Field::UByte(val) => Some(*val),
        Field::Null => None,
        field => bail!("Unexpected field: '{field}'"),
    };
    Ok(bytes)
}

fn decode_as_optional_u16(field: &Field) -> anyhow::Result<Option<u16>> {
    let val = match field {
        Field::UShort(val) => Some(*val),
        Field::Null => None,
        field => bail!("Unexpected field: '{field}'"),
    };
    Ok(val)
}

fn decode_as_optional_u64(field: &Field) -> anyhow::Result<Option<u64>> {
    let val = match field {
        Field::ULong(val) => Some(*val),
        Field::Null => None,
        field => bail!("Unexpected field: '{field}'"),
    };
    Ok(val)
}

fn decode_as_u64(field: &Field) -> anyhow::Result<u64> {
    decode_as_optional_u64(field)?.ok_or_else(|| anyhow!("Cannot be NULL"))
}

fn decode_as_optional_asset_id(field: &Field) -> anyhow::Result<Option<AssetId>> {
    Ok(decode_as_optional_bytes_32(field)?.map(|bytes_32| AssetId::new(*bytes_32)))
}

fn decode_as_asset_id(field: &Field) -> anyhow::Result<AssetId> {
    decode_as_optional_asset_id(field)?.ok_or_else(|| anyhow!("Cannot be NULL"))
}

fn decode_as_bytes_32(field: &Field) -> anyhow::Result<Bytes32> {
    decode_as_optional_bytes_32(field)?.ok_or_else(|| anyhow!("Cannot be NULL"))
}

fn decode_as_optional_bytes_32(field: &Field) -> anyhow::Result<Option<Bytes32>> {
    let bytes = decode_as_optional_bytes(field)?;
    bytes
        .map(|bytes| -> anyhow::Result<_> {
            let data = bytes.try_into()?;
            Ok(Bytes32::new(data))
        })
        .transpose()
}

fn decode_as_optional_block_height(field: &Field) -> anyhow::Result<Option<BlockHeight>> {
    let val = match field {
        Field::UInt(val) => Some(*val),
        Field::Null => None,
        field => bail!("Unexpected field: '{field}'"),
    };
    Ok(val.map(BlockHeight::new))
}

fn decode_as_optional_address(field: &Field) -> anyhow::Result<Option<Address>> {
    Ok(decode_as_optional_bytes_32(field)?.map(|bytes_32| Address::new(*bytes_32)))
}

fn decode_as_address(field: &Field) -> anyhow::Result<Address> {
    decode_as_optional_address(field)?.ok_or_else(|| anyhow!("Cannot be NULL"))
}

impl TryFrom<Row> for MessageConfig {
    type Error = anyhow::Error;
    fn try_from(row: Row) -> Result<Self, Self::Error> {
        let mut iter = row.get_column_iter();
        let mut next_field = || {
            iter.next()
                .map(|el| el.1)
                .ok_or_else(|| anyhow!("Expected at least one more field. Row: {row:?}"))
        };

        let sender = next_field()
            .and_then(decode_as_address)
            .context("While decoding `sender`")?;

        let recipient = next_field()
            .and_then(decode_as_address)
            .context("While decoding `recipient`")?;

        let nonce = next_field()
            .and_then(decode_as_bytes_32)
            .map(|bytes_32| Nonce::new(*bytes_32))
            .context("While decoding 'nonce'")?;

        let amount = next_field()
            .and_then(decode_as_u64)
            .context("While decoding `amount`")?;

        let data = next_field()
            .and_then(decode_as_bytes)
            .context("While decoding `data`")?
            .to_vec();

        let da_height = next_field()
            .and_then(decode_as_u64)
            .map(DaBlockHeight)
            .context("While decoding `amount`")?;

        Ok(Self {
            sender,
            recipient,
            nonce,
            amount,
            data,
            da_height,
        })
    }
}

impl TryFrom<Row> for ContractStateConfig {
    type Error = anyhow::Error;
    fn try_from(row: Row) -> Result<Self, Self::Error> {
        let mut iter = row.get_column_iter();
        let mut next_field = || {
            iter.next()
                .map(|el| el.1)
                .ok_or_else(|| anyhow!("Expected at least one more field. Row: {row:?}"))
        };

        let contract_id = next_field()
            .and_then(decode_as_bytes_32)
            .context("While decoding `contract_id`")?;

        let key = next_field()
            .and_then(decode_as_bytes_32)
            .context("While decoding `key`")?;

        let value = next_field()
            .and_then(decode_as_bytes_32)
            .context("While decoding `value`")?;

        Ok(Self {
            contract_id,
            key,
            value,
        })
    }
}

impl TryFrom<Row> for ContractBalanceConfig {
    type Error = anyhow::Error;
    fn try_from(row: Row) -> Result<Self, Self::Error> {
        let mut iter = row.get_column_iter();
        let mut next_field = || {
            iter.next()
                .map(|el| el.1)
                .ok_or_else(|| anyhow!("Expected at least one more field. Row: {row:?}"))
        };

        let contract_id = next_field()
            .and_then(decode_as_bytes_32)
            .context("While decoding `contract_id`")?;

        let asset_id = next_field()
            .and_then(decode_as_bytes_32)
            .map(|bytes_32| AssetId::new(*bytes_32))
            .context("While decoding `contract_id`")?;

        let amount = next_field()
            .and_then(decode_as_u64)
            .context("While decoding `amount`")?;

        Ok(Self {
            contract_id,
            asset_id,
            amount,
        })
    }
}

impl TryFrom<Row> for ContractConfig {
    type Error = anyhow::Error;
    fn try_from(row: Row) -> Result<Self, Self::Error> {
        let mut iter = row.get_column_iter();
        let mut next_field = || {
            iter.next()
                .map(|el| el.1)
                .ok_or_else(|| anyhow!("Expected at least one more field. Row: {row:?}"))
        };

        let contract_id = next_field()
            .and_then(decode_as_bytes_32)
            .map(|bytes_32| ContractId::new(*bytes_32))
            .context("While decoding `contract_id`")?;

        let code = next_field()
            .and_then(decode_as_bytes)
            .context("While decoding `code`")?
            .to_vec();

        let salt = next_field()
            .and_then(decode_as_bytes_32)
            .map(|bytes_32| Salt::new(*bytes_32))
            .context("While decoding `salt`")?;

        let tx_id = next_field()
            .and_then(decode_as_optional_bytes_32)
            .context("While decoding `tx_id`")?;

        let output_index = next_field()
            .and_then(decode_as_optional_u8)
            .context("While decoding `output_index`")?;

        let tx_pointer_block_height = next_field()
            .and_then(decode_as_optional_block_height)
            .context("While decoding `tx_pointer_block_height`")?;

        let tx_pointer_tx_idx = next_field()
            .and_then(decode_as_optional_u16)
            .context("While decoding `tx_pointer_tx_idx`")?;

        Ok(Self {
            contract_id,
            code,
            salt,
            tx_id,
            output_index,
            tx_pointer_block_height,
            tx_pointer_tx_idx,
            state: None,
            balances: None,
        })
    }
}
