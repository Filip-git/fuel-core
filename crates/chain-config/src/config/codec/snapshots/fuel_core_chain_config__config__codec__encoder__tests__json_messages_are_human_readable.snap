---
source: crates/chain-config/src/config/codec/encoder.rs
expression: encoded_json
---
{
  "coins": [],
  "messages": [
    {
      "sender": "0x0101010101010101010101010101010101010101010101010101010101010101",
      "recipient": "0x0202020202020202020202020202020202020202020202020202020202020202",
      "nonce": "0x0303030303030303030303030303030303030303030303030303030303030303",
      "amount": 4,
      "data": "0x0505050505050505050505050505050505050505050505050505050505050505",
      "da_height": 6
    }
  ],
  "contracts": [],
  "contract_state": [],
  "contract_balance": []
}
